### ISSUE TITLE
Lokr training shows a geometric increase in training time.

### ISSUE THREAD

**Issue by w844741652** [ostris/ai-toolkit#595]
Skipping first sample due to config setting
lokr-test-2:  10%|9         | 299/3000 [10:44<1:37:05,  2.16s/it, lr: 2.0e-04 loss: 5.226e-01]
Saving at step 300
Saved checkpoint to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000000300.safetensors
Saved optimizer to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\optimizer.pt
lokr-test-2:  20%|#9        | 599/3000 [19:45<1:03:24,  1.58s/it, lr: 2.0e-04 loss: 6.736e-01]
Saving at step 600
Saved checkpoint to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000000600.safetensors
Saved optimizer to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\optimizer.pt
lokr-test-2:  30%|##9       | 899/3000 [29:51<1:06:28,  1.90s/it, lr: 2.0e-04 loss: 6.463e-01]
Saving at step 900
Saved checkpoint to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000000900.safetensors
Saved optimizer to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\optimizer.pt
lokr-test-2:  40%|###9      | 1199/3000 [2:49:04<14:04:18, 28.13s/it, lr: 2.0e-04 loss: 7.353e-01]
Saving at step 1200
Saved checkpoint to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000001200.safetensors
Saved optimizer to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\optimizer.pt
lokr-test-2:  50%|####9     | 1499/3000 [2:56:55<34:04,  1.36s/it, lr: 2.0e-04 loss: 2.635e-01]
Saving at step 1500
Saved checkpoint to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000001500.safetensors
Saved optimizer to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\optimizer.pt
Removing old save: D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000000300.safetensors
lokr-test-2:  60%|#####9    | 1799/3000 [3:04:52<28:50,  1.44s/it, lr: 2.0e-04 loss: 6.560e-01]
Saving at step 1800
Saved checkpoint to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000001800.safetensors
Saved optimizer to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\optimizer.pt
Removing old save: D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000000600.safetensors
lokr-test-2:  70%|######9   | 2099/3000 [3:12:42<19:30,  1.30s/it, lr: 2.0e-04 loss: 5.365e-01]
Saving at step 2100
Saved checkpoint to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000002100.safetensors
Saved optimizer to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\optimizer.pt
Removing old save: D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000000900.safetensors
lokr-test-2:  80%|#######9  | 2399/3000 [3:21:06<18:15,  1.82s/it, lr: 2.0e-04 loss: 7.797e-01]
Saving at step 2400
Saved checkpoint to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000002400.safetensors
Saved optimizer to D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\optimizer.pt
Removing old save: D:\AI-Toolkit\AI-Toolkit\output\lokr-test-2\lokr-test-2_000001200.safetensors
lokr-test-2:  89%|########8 | 2656/3000 [5:10:12<3:11:06, 33.33s/it, lr: 2.0e-04 loss: 2.033e-01]

**Comment by ChaoZhao4056** [comment]
I also encountered this problem when I trained ZIT lora, the training was getting slower and slower with each iteration. Everything was normal when I trained the first lora last night. I kept encountering this problem when training lora today.

**Comment by ChaoZhao4056** [comment]
I fixed this problem with Gemini3pro, CPU memory leak fix is ​​not necessary, the key is GPU memory leak.

> # Technical Report on Training Slowdown Fixes
>
> ## Problem Description
> Users observed that training speed progressively slowed down as steps increased when training LoRA (specifically with Z-Image Turbo models) using AI Toolkit. Analysis revealed two main causes:
> 1.  **CPU Memory Leak**: In the WebUI trainer, asynchronous task objects used for status updates were accumulating indefinitely, causing increased memory usage and GC pressure.
> 2.  **GPU VRAM Accumulation/Fragmentation**: Under high load, PyTorch's caching allocator was not releasing VRAM timely, causing spillover into system shared memory, which severely impacted performance.
>
> ## Fix Details
>
> ### 1. Fix Asynchronous Task Memory Leak
> **File**: `extensions_built_in/sd_trainer/DiffusionTrainer.py`
>
> **Purpose**: Prevent the `self._async_tasks` list from growing indefinitely during training.
>
> **Modification**:
> In the `_run_async_operation` method:
> -   For synchronously executed tasks (`loop.run_until_complete`), stopped adding them to the `_async_tasks` list since they are completed upon return.
> -   For asynchronously executed tasks, added simple cleanup logic to remove completed tasks.
>
> ```python
> # Before
> if loop.is_running():
>     task = asyncio.run_coroutine_threadsafe(coro, loop)
>     self._async_tasks.append(asyncio.wrap_future(task))
> else:
>     task = loop.create_task(coro)
>     self._async_tasks.append(task)
>     loop.run_until_complete(task)
>
> # After
> if loop.is_running():
>     task = asyncio.run_coroutine_threadsafe(coro, loop)
>     self._async_tasks.append(asyncio.wrap_future(task))
>     # Clean up finished tasks to prevent memory leak
>     self._async_tasks = [t for t in self._async_tasks if not t.done()]
> else:
>     task = loop.create_task(coro)
>     # Sync execution, no need to track since it is done when we return
>     loop.run_until_complete(task)
> ```
>
> ### 2. Force VRAM Cleanup
> **File**: `extensions_built_in/sd_trainer/SDTrainer.py`
>
> **Purpose**: Force GPU VRAM cleanup and Python garbage collection after each training step to prevent VRAM fragmentation and spillover to shared memory.
>
> **Modification**:
> Added a `flush()` call before returning in the `hook_train_loop` method.
>
> ```python
> # Before
> self.end_of_training_loop()
> return loss_dict
>
> # After
> self.end_of_training_loop()
> flush() # Force VRAM cleanup
> return loss_dict
> ```

**Comment by update-loulan** [comment]
@ChaoZhao4056 感谢老哥，我也去问问gemini改改脚本

**Comment by Illynir** [comment]
I had the same bug and the fix of ChaosZhao worked perfectly.

LOKR was untrainable because of that, now yes. thanks.
